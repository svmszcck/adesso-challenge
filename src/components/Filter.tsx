import React, { useState } from "react";
import { View, TouchableOpacity, StyleSheet } from "react-native";
import Ionicons from "@expo/vector-icons/Ionicons";

import { useThemeColor } from "@/hooks/useThemeColor";
import { movieTypes, FontSize, Spacing } from "@/constants";
import ListModal from "@/utils/modals/ListModal";
import { Button, ThemedText } from "@/components";
import { MovieType } from "@/services/types";
import { generateYearRange, normalizeListData } from "@/utils/data";

type FilterProps = {
  onClear?: () => void;
  type: string | undefined;
  setType: (value: MovieType) => void;
  year: string | undefined;
  setYear: (value: string) => void;
};

const Filter: React.FC<FilterProps> = ({
  type,
  setType,
  year,
  setYear,
  onClear,
}) => {
  const textPaleColor = useThemeColor({}, "textPale");
  const [typeModal, setTypeModal] = useState<boolean>(false);
  const [yearModal, setYearModal] = useState<boolean>(false);

  const onSelectType = (value: string) => {
    setType(value as MovieType);
    setTypeModal(false);
  };

  const onSelectYear = (value: string) => {
    setYear(value);
    setYearModal(false);
  };

  return (
    <View style={styles.container}>
      <TouchableOpacity
        style={styles.option}
        activeOpacity={0.5}
        onPress={() => setTypeModal(true)}
      >
        <ThemedText type="default" style={styles.optionText} isPale>
          Tür
        </ThemedText>
        <Ionicons name="chevron-down" size={20} color={textPaleColor} />
      </TouchableOpacity>

      <TouchableOpacity
        style={styles.option}
        activeOpacity={0.5}
        onPress={() => setYearModal(true)}
      >
        <ThemedText type="default" style={styles.optionText} isPale>
          Yayınlanma Yılı
        </ThemedText>
        <Ionicons name="chevron-down" size={20} color={textPaleColor} />
      </TouchableOpacity>

      {onClear && (type || year) && (
        <Button title="Sıfırla" onPress={onClear} isSmall />
      )}

      <ListModal
        visible={typeModal}
        title="Tür Seçerek Filtrele"
        data={movieTypes}
        onClose={() => setTypeModal(false)}
        onSelect={onSelectType}
        selected={type}
      />
      <ListModal
        visible={yearModal}
        title="Yıl Seçerek Filtrele"
        data={normalizeListData(generateYearRange(1950))}
        onClose={() => setYearModal(false)}
        onSelect={onSelectYear}
        selected={year}
      />
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    flexDirection: "row",
    marginBottom: Spacing.SMALL,
  },
  option: {
    flexDirection: "row",
    alignItems: "center",
    marginRight: Spacing.MEDIUM,
  },
  optionText: {
    marginRight: 5,
    lineHeight: 20,
  },
  contentTitle: {
    fontSize: FontSize.LARGE,
    textAlign: "center",
  },
  modalTitle: {
    marginBottom: 50,
  },
  modalListItem: {
    width: "100%",
  },
  separator: {
    width: "100%",
    height: 1,
    marginVertical: Spacing.LARGE,
  },
});

export default Filter;
