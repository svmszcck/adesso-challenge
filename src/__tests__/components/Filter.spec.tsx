import React from "react";
import { render, fireEvent, screen } from "@testing-library/react-native";
import Filter from "@/components/Filter";

describe("Filter Component", () => {
  const mockSetType = jest.fn();
  const mockSetYear = jest.fn();
  const mockOnClear = jest.fn();

  const defaultProps = {
    type: undefined,
    setType: mockSetType,
    year: undefined,
    setYear: mockSetYear,
    onClear: mockOnClear,
  };

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it("renders correctly", () => {
    render(<Filter {...defaultProps} />);

    // Check that buttons and icons are displayed
    expect(screen.getByText("Tür")).toBeTruthy();
    expect(screen.getByText("Yayınlanma Yılı")).toBeTruthy();
  });

  it("opens type modal when 'Tür' button is pressed", () => {
    render(<Filter {...defaultProps} />);

    // Open type modal
    fireEvent.press(screen.getByText("Tür"));

    // Check if type modal opens
    expect(screen.getByText("Tür Seçerek Filtrele")).toBeTruthy();
  });

  it("opens year modal when 'Yayınlanma Yılı' button is pressed", () => {
    render(<Filter {...defaultProps} />);

    // Open year modal
    fireEvent.press(screen.getByText("Yayınlanma Yılı"));

    // Check if year modal opens
    expect(screen.getByText("Yıl Seçerek Filtrele")).toBeTruthy();
  });

  it("calls setType and closes type modal on selection", () => {
    render(<Filter {...defaultProps} />);

    // Open type modal and select a type
    fireEvent.press(screen.getByText("Tür"));
    fireEvent.press(screen.getByText("Film"));

    // Check if setType was called with the correct value
    expect(mockSetType).toHaveBeenCalledWith("movie");
  });

  it("calls setYear and closes year modal on selection", () => {
    render(<Filter {...defaultProps} />);

    // Open year modal and select a year
    fireEvent.press(screen.getByText("Yayınlanma Yılı"));
    fireEvent.press(screen.getByText("2022"));

    // Check if setYear was called with the correct value
    expect(mockSetYear).toHaveBeenCalledWith("2022");
  });

  it("renders clear button if type or year is set", () => {
    render(<Filter {...defaultProps} type="Action" />);

    // Check that the clear button is rendered
    expect(screen.getByText("Sıfırla")).toBeTruthy();
  });

  it("calls onClear when clear button is pressed", () => {
    render(<Filter {...defaultProps} type="Action" year="2022" />);

    // Press clear button
    fireEvent.press(screen.getByText("Sıfırla"));

    // Verify onClear callback
    expect(mockOnClear).toHaveBeenCalledTimes(1);
  });

  it("matches snapshot", () => {
    const { toJSON } = render(
      <Filter
        type="movie"
        year="2018"
        setType={mockSetType}
        setYear={mockSetYear}
        onClear={mockOnClear}
      />
    );
    expect(toJSON()).toMatchSnapshot();
  });
});
