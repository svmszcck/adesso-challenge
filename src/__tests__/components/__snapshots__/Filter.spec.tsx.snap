// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Filter Component matches snapshot 1`] = `
<View
  style={
    {
      "flexDirection": "row",
      "marginBottom": 10,
    }
  }
>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "flexDirection": "row",
        "marginRight": 15,
        "opacity": 1,
      }
    }
  >
    <Text
      style={
        [
          {
            "color": "#898999",
          },
          {
            "fontSize": 16,
            "lineHeight": 24,
          },
          undefined,
          undefined,
          undefined,
          undefined,
          {
            "lineHeight": 20,
            "marginRight": 5,
          },
        ]
      }
    >
      Tür
    </Text>
    <Text
      allowFontScaling={false}
      selectable={false}
      style={
        [
          {
            "color": "#898999",
            "fontSize": 20,
          },
          undefined,
          {
            "fontFamily": "ionicons",
            "fontStyle": "normal",
            "fontWeight": "normal",
          },
          {},
        ]
      }
    >
      
    </Text>
  </View>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "flexDirection": "row",
        "marginRight": 15,
        "opacity": 1,
      }
    }
  >
    <Text
      style={
        [
          {
            "color": "#898999",
          },
          {
            "fontSize": 16,
            "lineHeight": 24,
          },
          undefined,
          undefined,
          undefined,
          undefined,
          {
            "lineHeight": 20,
            "marginRight": 5,
          },
        ]
      }
    >
      Yayınlanma Yılı
    </Text>
    <Text
      allowFontScaling={false}
      selectable={false}
      style={
        [
          {
            "color": "#898999",
            "fontSize": 20,
          },
          undefined,
          {
            "fontFamily": "ionicons",
            "fontStyle": "normal",
            "fontWeight": "normal",
          },
          {},
        ]
      }
    >
      
    </Text>
  </View>
  <View
    accessibilityLabel="Sıfırla"
    accessibilityRole="button"
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "backgroundColor": "#364249",
        "borderRadius": 10,
        "elevation": 8,
        "opacity": 1,
        "paddingHorizontal": 12,
        "paddingVertical": 6,
      }
    }
  >
    <Text
      style={
        [
          {
            "color": "#364249",
          },
          {
            "fontSize": 16,
            "lineHeight": 24,
          },
          undefined,
          undefined,
          undefined,
          undefined,
          [
            {
              "alignSelf": "center",
              "fontSize": 18,
              "fontWeight": "bold",
            },
            {
              "color": "#ffffff",
            },
          ],
        ]
      }
    >
      Sıfırla
    </Text>
  </View>
</View>
`;
